### PR TITLE
unpack: optimize unpack.Tgz memory allocations

### PR DESCRIPTION
**Pull request by Strum355**
In the worst case (when the `io.Reader` isnt an `io.ReadSeeker`), `unpack.Tgz` would use `io.ReadAll` on the `io.Reader` in order to fall-back to `unpack.Tar` if the file extension was incorrect (.tar.gz for something what wasnt actually gzipped).  

After https://github.com/sourcegraph/sourcegraph/pull/44497/, https://github.com/sourcegraph/sourcegraph/pull/44517/ and https://github.com/sourcegraph/sourcegraph/pull/44520, this worst will always be triggered, as http response bodies are not `io.ReadSeeker`s. This can cause excessive memory copying from `io.ReadAll`, where we could be reading a minimal amount to determine whether the file is actually gzipped or not.

Benchmark figures, where Old is the original, and New is the new (of course):
```
$ go test -benchmem -run='^$' -bench '^BenchmarkTgzFallback$' github.com/sourcegraph/sourcegraph/internal/unpack -v -count=10 -race
goos: linux
goarch: amd64
pkg: github.com/sourcegraph/sourcegraph/internal/unpack
cpu: AMD Ryzen 5 5600X 6-Core Processor             
BenchmarkTgzFallback
BenchmarkTgzFallback/Old
BenchmarkTgzFallback/Old-12                29144             43833 ns/op           14557 B/op         36 allocs/op
BenchmarkTgzFallback/Old-12                26395             45004 ns/op           14574 B/op         36 allocs/op
BenchmarkTgzFallback/Old-12                28130             45917 ns/op           14599 B/op         36 allocs/op
BenchmarkTgzFallback/Old-12                26758             46788 ns/op           14564 B/op         36 allocs/op
BenchmarkTgzFallback/Old-12                25844             45694 ns/op           14532 B/op         36 allocs/op
BenchmarkTgzFallback/Old-12                27319             45318 ns/op           14512 B/op         36 allocs/op
BenchmarkTgzFallback/Old-12                27026             46731 ns/op           14586 B/op         36 allocs/op
BenchmarkTgzFallback/Old-12                26932             47716 ns/op           14538 B/op         36 allocs/op
BenchmarkTgzFallback/Old-12                25428             45836 ns/op           14563 B/op         36 allocs/op
BenchmarkTgzFallback/Old-12                26865             41620 ns/op           14530 B/op         36 allocs/op
BenchmarkTgzFallback/New
BenchmarkTgzFallback/New-12                33439             40709 ns/op            5945 B/op         32 allocs/op
BenchmarkTgzFallback/New-12                29628             39881 ns/op            5940 B/op         31 allocs/op
BenchmarkTgzFallback/New-12                31644             39578 ns/op            5971 B/op         32 allocs/op
BenchmarkTgzFallback/New-12                29116             39691 ns/op            5957 B/op         32 allocs/op
BenchmarkTgzFallback/New-12                30648             39306 ns/op            5995 B/op         32 allocs/op
BenchmarkTgzFallback/New-12                31161             41873 ns/op            5984 B/op         32 allocs/op
BenchmarkTgzFallback/New-12                30362             41084 ns/op            5939 B/op         32 allocs/op
BenchmarkTgzFallback/New-12                30147             39431 ns/op            5999 B/op         32 allocs/op
BenchmarkTgzFallback/New-12                31102             39300 ns/op            5933 B/op         32 allocs/op
BenchmarkTgzFallback/New-12                32269             42896 ns/op            5977 B/op         32 allocs/op
```

## Test plan

Unit test, benchmark, code review, live :)

### CODE REVIEW

File: internal/unpack/unpack.go [hunk 1]
```diff
@@ -181,9 +180,9 @@ func extractTarFile(tr *tar.Reader, h *tar.Header, dir string) error {
 
 	// We need to be able to traverse directories and read/modify files.
 	if mode.IsDir() {
-		mode |= 0700
+		mode |= 0o700
```
Comment (eseliger) [inline internal/unpack/unpack.go:183]: this is pretty unrelated, can we change it back to keep the diff clean?

File: internal/unpack/bi_reader.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+// biReader is a specialized io.MultiReader optimized for
+// only two readers
+type biReader struct {
+	readers [2]io.Reader
+	cur     uint8
+}
+
+func (mr *biReader) Read(p []byte) (n int, err error) {
+	cur := mr.cur
```
Comment (eseliger) [inline internal/unpack/bi_reader.go:13]: can you add a comment that this line is here intentionally?

File: internal/unpack/unpack.go (hunk 1 above)
Comment (Strum355) [reply internal/unpack/unpack.go:183]: I do as gopls commands 🤖 its automatic, but ya ill get rid of it

File: internal/unpack/unpack_test.go [hunk 3]
```diff
@@ -20,7 +20,7 @@ import (
 )
 
 func TestTgzFallback(t *testing.T) {
```
Comment (eseliger) [inline internal/unpack/unpack_test.go:22]: looks like all the changes in here could be reverted?

**Review by eseliger — APPROVED** [review]
cool stuff!

File: internal/unpack/bi_reader.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	if cur < 2 {
+		n, err = mr.readers[cur].Read(p)
+		if err == io.EOF {
+			mr.readers[cur] = nil
+			mr.cur++
+		}
+		if mr.cur == 1 && err == io.EOF {
+			err = nil
+		}
+	}
```
Comment (efritz) [inline internal/unpack/bi_reader.go:24]:
```suggestion
	if cur >= 2 {
		return 0, io.EOF
	}
	n, err = mr.readers[cur].Read(p)
	if err == io.EOF {
		mr.readers[cur] = nil
		mr.cur++
		if mr.cur == 1 {
			err = nil
		}
	}
	return
}
```

Don't you need to ensure that `io.EOF` is returned on all subsequent calls?

Comment (Strum355) [reply internal/unpack/bi_reader.go:24]: Probably yes lol, this was specifically the area I wanted eyes on coz its so damn fiddly

File: internal/unpack/unpack_test.go (hunk 3 above)
Comment (Strum355) [reply internal/unpack/unpack_test.go:22]: oop ya, theres another one

File: internal/unpack/unpack.go [hunk 5]
```diff
@@ -77,37 +77,35 @@ func Zip(r io.ReaderAt, size int64, dir string, opt Opts) error {
...
-	// from gzip.NewReader() trying to read a gzip header, so make
-	// sure we have an io.ReadSeeker.
-	tgz, ok := r.(io.ReadSeeker)
-	if !ok {
-		bs, err := io.ReadAll(r)
-		if err != nil {
-			return err
-		}
-		tgz = bytes.NewReader(bs)
+	var gzipMagicBytes [2]byte
```
[the remark this answers is not in the document]
Comment (Strum355) [reply internal/unpack/unpack.go:94]: [`302020d` (#44529)](https://github.com/sourcegraph/sourcegraph/pull/44529/commits/302020dcf220a92a1e11a593e19706decbba1a41)

**Review by eseliger — APPROVED** [review]

File: internal/unpack/bi_reader.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	second io.Reader
+}
+
+func (mr *biReader) Read(p []byte) (n int, err error) {
+	if mr.first != nil {
+		n, err = mr.first.Read(p)
+		if err == io.EOF {
+			err = nil
+			mr.first = nil
+		}
```
Comment (keegancsmith) [inline internal/unpack/bi_reader.go:18]: Quoting the io.Reader docs, this can happen.

> Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0.

File: internal/unpack/bi_reader.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package unpack
+
+import "io"
+
+// biReader is a specialized io.MultiReader optimized for
+// only two readers
+type biReader struct {
```
Comment (keegancsmith) [inline internal/unpack/bi_reader.go:7]: This is really cool. Out of interest, how much of a difference did it make? Worth trying to upstream the specialization to MultiReader?

File: internal/unpack/bi_reader.go (hunk 6 above)
Comment (Strum355) [reply internal/unpack/bi_reader.go:18]: Im assuming youre referencing a case where `m.first.Read` returns 0 and EOF? Thats a good catch, the sentence that follows should mean that we're still going to be sound in theory
> Callers should treat a return of 0 and nil as indicating that nothing happened; in particular it does not indicate EOF.
